src/mpl2/src/clusterEngine.cpp

```diff
@@ -1432,33 +1432,39 @@ void ClusteringEngine::breakLargeFlatCluster(Cluster* parent)
   }
 
   const int seed = 0;
-  const float balance_constraint = 1.0;
+  constexpr float default_balance_constraint = 1.0f;
+  float balance_constraint = default_balance_constraint;
   const int num_parts = 2;  // We use two-way partitioning here
   const int num_vertices = static_cast<int>(vertex_weight.size());
   std::vector<float> hyperedge_weights(hyperedges.size(), 1.0f);
 
-  debugPrint(logger_,
-             MPL,
-             "multilevel_autoclustering",
-             1,
-             "Breaking flat cluster {} with TritonPart",
-             parent->getName());
-
-  std::vector<int> part
-      = triton_part_->PartitionKWaySimpleMode(num_parts,
-                                              balance_constraint,
-                                              seed,
-                                              hyperedges,
-                                              vertex_weight,
-                                              hyperedge_weights);
-
-  if (partitionerSolutionIsFullyUnbalanced(part, num_other_cluster_vertices)) {
-    logger_->error(MPL,
-                   37,
-                   "Couldn't break flat cluster {} with PAR. The solution is "
-                   "fully unbalanced.",
-                   parent->getName());
-  }
+  // Due to the discrepancy that may exist between the weight of vertices
+  // that represent macros/std cells, the partitioner may fail to meet the
+  // balance constraint. This may cause the output to be completely unbalanced
+  // and lead to infinite partitioning recursion. To handle that, we relax
+  // the constraint until we find a reasonable split.
+  constexpr float balance_constraint_relaxation_factor = 10.0f;
+  std::vector<int> solution;
+  do {
+    debugPrint(
+        logger_,
+        MPL,
+        "multilevel_autoclustering",
+        1,
+        "Attempting flat cluster {} partitioning with balance constraint = {}",
+        parent->getName(),
+        balance_constraint);
+
+    solution = triton_part_->PartitionKWaySimpleMode(num_parts,
+                                                     balance_constraint,
+                                                     seed,
+                                                     hyperedges,
+                                                     vertex_weight,
+                                                     hyperedge_weights);
+
+    balance_constraint += balance_constraint_relaxation_factor;
+  } while (partitionerSolutionIsFullyUnbalanced(solution,
+                                                num_other_cluster_vertices));
```

Comment:
couldn't this go on forever?

Reply:
I think it may be possible to come up with a case in which it will go on for a very long time. E.g., a cluster with 1 macro and 3 std cells for which we set the max std cell threshold to be 2. It would be quite extreme though.

However if we want some defensive measure we can limit the amount of tries? I'm open to suggestions.

Reply:
balance_constraint  will eventually become too large to have much meaning.  Probably we should stop at 90%